prometheus_juju_exporter/collector.py

```diff
@@ -47,21 +48,32 @@ def refresh_cache(self, gauge_name: str, gauge_desc: str, labels: List[str]) ->
         self.controller = Controller(max_frame_size=6**24)
 
     async def _connect_controller(
-        self, endpoint: str, username: str, password: str, cacert: str
+        self, endpoints: List[str], username: str, password: str, cacert: str
     ) -> None:
         """Connect to a controller via its endpoint.
 
-        :param str endpoint: the hostname:port endpoint of the controller
+        :param List[str] endpoints: list of the hostname:port endpoints of the controllers
             to connect to.
         :param str username: the username for controller-local users
         :param str password: the password for controller-local users
         :param str cacert: the CA certificate of the controller
             (PEM formatted)
         """
         if not self.controller.is_connected():
-            await self.controller.connect(
-                endpoint=endpoint, username=username, password=password, cacert=cacert
-            )
+            for endpoint in endpoints:
+                self.logger.info("Connecting to controller at %s", endpoint)
+                try:
+                    await self.controller.connect(
+                        endpoint=endpoint, username=username, password=password, cacert=cacert
+                    )
+                    break
+                except JujuError as exc:
```

Comment:
Are you sure that catching only JujuError is enough here? libjuju's `connection.connect()` have a catch-all exception (https://github.com/juju/python-libjuju/blob/master/juju/client/connection.py#L358) implemented to handle all types of connection failures. Since we are doing fail-over, if connecting to one of the endpoint resulting in such an exception, I think we should catch it without terminating the process.